xmtp_api_grpc/src/replication_client.rs

```diff
@@ -170,207 +160,166 @@ impl XmtpApiClient for ClientV4 {
+                        Error::new(ErrorKind::InternalError(InternalError::MissingPayloadError))
+                            .with("Missing key package")
+                    })?;
+
+                    Ok(fetch_key_packages_response::KeyPackage {
+                        key_package_tls_serialized: key_package.key_package_tls_serialized,
+                    })
+                } else {
+                    Err(
+                        Error::new(ErrorKind::InternalError(InternalError::MissingPayloadError))
+                            .with("Payload is not a key package"),
+                    )
+                }
+            })
+            .collect();
+
+        Ok(FetchKeyPackagesResponse {
+            key_packages: key_packages?,
+        })
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn send_group_messages(&self, req: SendGroupMessagesRequest) -> Result<(), Error> {
-        unimplemented!();
+        self.publish_envelopes_to_payer(req.messages).await
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn send_welcome_messages(&self, req: SendWelcomeMessagesRequest) -> Result<(), Error> {
-        unimplemented!();
+        self.publish_envelopes_to_payer(req.messages).await
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn query_group_messages(
         &self,
         req: QueryGroupMessagesRequest,
     ) -> Result<QueryGroupMessagesResponse, Error> {
-        unimplemented!();
+        let client = &mut self.client.clone();
+        let res = client
+            .query_envelopes(QueryEnvelopesRequest {
+                query: Some(EnvelopesQuery {
+                    topics: vec![build_group_message_topic(req.group_id.as_slice())],
+                    originator_node_ids: vec![],
+                    last_seen: None,
+                }),
+                limit: 100,
+            })
+            .await
+            .map_err(|e| Error::new(ErrorKind::MlsError).with(e))?;
+
+        let envelopes = res.into_inner().envelopes;
+        let response = QueryGroupMessagesResponse {
+            messages: envelopes
+                .iter()
+                .map(|envelope| {
+                    let unsigned_originator_envelope =
+                        extract_unsigned_originator_envelope(envelope).unwrap();
+                    let client_envelope = extract_client_envelope(envelope).unwrap();
+                    let Payload::GroupMessage(group_message) = client_envelope.payload.unwrap()
+                    else {
+                        panic!("Payload is not a group message");
+                    };
+
+                    let group_message_input::Version::V1(v1_group_message) =
+                        group_message.version.unwrap();
+
+                    GroupMessage {
+                        version: Some(group_message::Version::V1(group_message::V1 {
+                            id: unsigned_originator_envelope.originator_sequence_id,
```

Comment:
Just making a note, eventually we'll need to refactor the client to store both the originator node id and sequence ID in order to set the vector clock properly, but this is great for now